numpy/testing/tests/test_utils.py

```diff
@@ -1123,7 +1123,15 @@ def assert_warn_len_equal(mod, n_in_context, py3_n_in_context=None):
     if 'version' in mod_warns:
         if py3_n_in_context is None:
             py3_n_in_context = n_in_context
-        assert_equal(len(mod_warns) - 1, py3_n_in_context)
+        if sys.version_info[:2] < (3, 7):
+            assert_equal(len(mod_warns) - 1, py3_n_in_context)
+        else:
+            if other and len(mod_warns) == 1:
+                # The behavior of warnings changed in Python 3.7 to be more
+                # module specific.
```

Comment:
I don't understand the 3.7 behavior here. Can you comment the `-1` case too?

Reply:
AFAICT, the old Python3 behavior was to make an entry in the `__warningregistry__` in all the modules up the line from the module in which the warning was issued, in 3.7 it looks like it is only in the module in which raised, hence the distinction of `other`. The `-1` is generic to Python 3, as a version entry is added to all the created registries. Note that some of our functions clear the registries all together, which really isn't quite right for Python3. The tests are a bit of a kludge, and hard to maintain as the tested functions fiddle with private Python matters, but they do seem to work as advertized. These aren't my functions though, maybe @seberg can add more.

Reply:
Yeah, the module warning filter is  a bit strange (and possibly not as reliable as it could be). The bigger user is the `clear_and_catch_warnings` which I do not think we use anymore likely (so the behaviour tested is likely used in a single place and should not even matter, luckily).

The -1 is of course due to the "version" entry which was added in 3.6 and allows picking up changes in the filter list, which used to be very broken.

The change in py 3.7 is a bit unclear to me right now (I really should run a python3.7). I would expect something different going on, possibly:
  1. >=Py3.6 adds the "version" information (to get rid of old activated filters
  2. This was broken in the sense of breaking "once" filters to possibly occur more then once, because the `__warningregistry__` was completely cleared!
  3. So Py3.7 fixes it (somewhat, I expect it can't be gotten 100% right really), so that "once" or "ignore" filters will stay in the `__warningregistry__` if there is still a matching filter active (which is the case here).

I expect "other" might want need to use the py2 behaviour here (not quite right probably too). I expect fully right would be something like:

  a) "number of old outdated filters"
  b) "number of old already active filters"
  c) "number of new filters"

Then Py2, will be `(a + b + c)`, py3.6 ("version", but not 3.7) will be just `c` and py3.7 will be `b+c`, add `-1` if "version" exists (py >3.6).

I have not checked it because I have no py3 install right now.

Reply:
Sorry, b) would be: Number of old *still* (apparently) active filters

Reply:
We could make `other` an `n_still_active` or so, and `a` number of all old filters so to not change the values in old py versions. The `n_py3...` is actually `c`, the `n_in_context` is actually `a+c`. Could also add the parameter: `number of non-surviving old filters=0`, since I bet 0 might work in all current places?

Reply:
Ok, I hope this was understandable, if someone wants a quick test:

```python

import warnings

# Distinguish python 2 from python 3.7:
# This warning should remain in the __warningregistry__ in python <3.6, but not in python 3.7 (or 3.6)
warnings.simplefilter('ignore', RuntimeWarning)
warnings.warn('A warning we will not ignore in the future!', RuntimeWarning)
warnings.simplefilter('always', RuntimeWarning)  # The last warning should be shown again (in principle)

print('Should have 1 warning (for all python versions):')
print('__warningregistry__')
print('-------------------')

warnings.warn("Hahahah!")  # Warning shown once only
print('< Py3.6 should have two filters (the old one survives, even though it is gone in principle), Py 3.6 and 3.7 should have one filter:')
print('__warningregistry__') 
print('-------------------')

# "Should" do nothing, but (partially) invalidates `__warningregistry__`
warnings.simplefilter('ignore', RuntimeWarning)  

warnings.warn("Now <Py 3.6 should have 3 filters `a+b+c`, Py=3.6 should have 1 (`c`) and Py3.7 should 2 (`b+c`)")
print('__warningregistry__')
```

Reply:
Tricky stuff… I think that for Python 3.5+, as long as you make sure to never have active "once" filters, testing is actually pretty warning clean even without `suppress_warnings`, but oh well, it has a bit of other niceties to offer additionally. Maybe at some point we only support those and pytest might add some nice things and we can dump it again.